File: frontend/lib/chart-spec.ts
import { z } from "zod";

// 与后端 Java record 双端对齐（05 §2.1/§2.2）。本期只含页面在用的三种类型；
// candlestick/table 随聊天流阶段扩展。null = 数据缺口（估值序列缺失日）。
const seriesData = z.array(z.union([z.number(), z.null()]));

export const ChartSpecSchema = z.discriminatedUnion("type", [
  z.object({
    specVersion: z.literal(1),
    type: z.literal("pie"),
    title: z.string(),
    subtitle: z.string().optional(),
    data: z.array(z.object({ name: z.string(), value: z.number() })),
    unit: z.string().optional(),
  }),
  z.object({
    specVersion: z.literal(1),
    type: z.literal("bar"),
    title: z.string(),
    subtitle: z.string().optional(),
    categories: z.array(z.string()),
    series: z.array(z.object({ name: z.string(), data: seriesData })),
    horizontal: z.boolean().optional(),
    unit: z.string().optional(),
  }),
  z.object({
    specVersion: z.literal(1),
    type: z.literal("line"),
    title: z.string(),
    subtitle: z.string().optional(),
    categories: z.array(z.string()),
    series: z.array(
      z.object({
        name: z.string(),
        data: seriesData,
        area: z.boolean().optional(),
      }),
    ),
    unit: z.string().optional(),
  }),
  z.object({
    specVersion: z.literal(1),
    type: z.literal("candlestick"),
    title: z.string(),
    subtitle: z.string().optional(),
    symbol: z.string(),                                  // "600519 贵州茅台"
    period: z.string(),                                  // "day"|"week"|"month"（与工具入参一致，展示用）
    dates: z.array(z.string()),                          // 交易日
    klines: z.array(z.tuple([z.number(), z.number(), z.number(), z.number()])),
    //                        ⚠ 顺序 = [开, 收, 低, 高]（ECharts 官方约定）
    volumes: z.array(z.number()).optional(),             // 成交量副图
    mas: z.array(z.object({
      name: z.string(),
      data: z.array(z.union([z.number(), z.null()])),    // null = 预热缺口，与 dates 对齐
    })).optional(),                                      // MA 线后端算好
  }),
  z.object({
    specVersion: z.literal(1),
    type: z.literal("table"),
    title: z.string(),
    subtitle: z.string().optional(),
    columns: z.array(z.object({
      key: z.string(), label: z.string(),
      align: z.enum(["left", "right", "center"]).optional(),
      sortable: z.boolean().optional(),                  // 默认 true
    })).min(1),                                          // 空列拒绝（后端契约保证 ≥1 列）
    rows: z.array(z.record(z.string(), z.union([z.string(), z.number(), z.null()]))),
    defaultPageSize: z.number().optional(),              // 预留：卡片内表格可视高度（本期固定 360px 未消费）
  }),
]);

export type ChartSpec = z.infer<typeof ChartSpecSchema>;
export type PieSpec = Extract<ChartSpec, { type: "pie" }>;
export type BarSpec = Extract<ChartSpec, { type: "bar" }>;
export type LineSpec = Extract<ChartSpec, { type: "line" }>;
export type CandlestickSpec = Extract<ChartSpec, { type: "candlestick" }>;
export type TableSpec = Extract<ChartSpec, { type: "table" }>;

File: backend/src/main/java/com/portfolio/invest/agent/chart/ChartSpec.java
package com.portfolio.invest.agent.chart;

import com.fasterxml.jackson.annotation.JsonInclude;
import java.util.List;
import java.util.Map;

/**
 * 与前端 lib/chart-spec.ts 对齐的判别联合（2026-09-11 澄清：单一契约，table 是变体之一）。
 * 显式携带 specVersion/type 组件（勿依赖 Jackson 类名推断）；每个 record 标 @JsonInclude(NON_NULL)：
 * wire 契约默认 ALWAYS 会把 null 字段发出去，而前端 zod .optional() 不接受 null。
 */
public sealed interface ChartSpec {
  int specVersion();
  String type();
  String title();
  String subtitle();

  @JsonInclude(JsonInclude.Include.NON_NULL)
  record Line(int specVersion, String type, String title, String subtitle,
              List<String> categories, List<Series> series, String unit) implements ChartSpec {}

  @JsonInclude(JsonInclude.Include.NON_NULL)
  record Series(String name, List<Double> data, Boolean area) {}   // data 允许 null 元素（缺口）

  @JsonInclude(JsonInclude.Include.NON_NULL)
  record Bar(int specVersion, String type, String title, String subtitle,
             List<String> categories, List<Series> series, Boolean horizontal, String unit)
          implements ChartSpec {}

  @JsonInclude(JsonInclude.Include.NON_NULL)
  record Pie(int specVersion, String type, String title, String subtitle,
             List<Slice> data, String unit) implements ChartSpec {}

  @JsonInclude(JsonInclude.Include.NON_NULL)
  record Slice(String name, Double value) {}

  /** klines 序列化为 4 元数值数组 [开,收,低,高]（前端 zod tuple，非对象）。 */
  @JsonInclude(JsonInclude.Include.NON_NULL)
  record Candlestick(int specVersion, String type, String title, String subtitle,
                     String symbol, String period, List<String> dates,
                     List<double[]> klines, List<Long> volumes, List<MaLine> mas)
          implements ChartSpec {}

  @JsonInclude(JsonInclude.Include.NON_NULL)
  record MaLine(String name, List<Double> data) {}                 // 预热期为 null 元素

  @JsonInclude(JsonInclude.Include.NON_NULL)
  record Table(int specVersion, String type, String title, String subtitle,
               List<Column> columns, List<Map<String, Object>> rows, Integer defaultPageSize)
          implements ChartSpec {}

  @JsonInclude(JsonInclude.Include.NON_NULL)
  record Column(String key, String label, String align, Boolean sortable) {}
}
